### PR TITLE
Link bookings to applications in Temporary Accommodation

### PR DESCRIPTION
**Pull request by danlivings-dxw**
> See [ticket #1381 on the CAS3 Trello board](https://trello.com/c/K6AjM2OT/1381-a-user-can-also-link-to-a-referral-as-part-of-creating-a-provisional-booking).

This PR introduces the ability to link bookings to applications within the Temporary Accommodation service. This is done by passing in an `assessmentId` parameter when sending a request to the `POST /premises/{premisesId}/bookings` endpoint.

This new field is optional, so the existing behaviour is preserved if this ID is not provided.

### CODE REVIEW

**Comment by bryangaledxw** [comment]
Tested UI manually against API. Looks good

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt [hunk 1]
```diff
@@ -540,6 +540,88 @@ class BookingTest : IntegrationTestBase() {
 
   @Test
   fun `Create Temporary Accommodation Booking returns OK with correct body`() {
+    `Given a User`(roles = listOf(UserRole.CAS3_ASSESSOR)) { userEntity, jwt ->
+      `Given an Offender` { offenderDetails, inmateDetails ->
+        val premises = temporaryAccommodationPremisesEntityFactory.produceAndPersist {
```
Comment (tahb) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt:545]: I wonder if there's a small opportunity to move this property+bed test set up into a helper function given it's not the focus of the test. That might then be reusable and make this and other tests like it more concise.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 2]
```diff
@@ -594,6 +597,17 @@ class BookingService(
         "$.departureDate" hasValidationError "beforeBookingArrivalDate"
       }
 
+      val application = when (assessmentId) {
+        null -> null
+        else -> {
+          val result = assessmentRepository.findByIdOrNull(assessmentId)
+          if (result == null) {
+            "$.assessmentId" hasValidationError "doesNotExist"
```
Comment (tahb) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:605]: It looks to set an error message on the object that's then validated in the next block which makes sense but I'm curious as to how this line works.

**Review by tahb — APPROVED** [review]
LGTM 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt (hunk 1 above)
Comment (tahb) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt:545]: I see now that this pattern of placing all set up in the test seems to be widely adopted already so any switch away should be more a coordinated effort.

Comment (danlivings-dxw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt:545]: I can put a ticket in the backlog to create a `Given a Premises`/`Given a Bed` helper function and make this switch?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt (hunk 2 above)
Comment (danlivings-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:605]: The `validated` method creates a [`ValidatedScope`](https://github.com/ministryofjustice/hmpps-approved-premises-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/ValidationErrors.kt), which holds a `MutableMap<String, String>` for validation errors.

The `hasValidationError` function is an [infix function](https://kotlinlang.org/docs/functions.html#infix-notation) which in this case puts `"$.assessmentId" -> "doesNotExist"` into the error map.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt (hunk 1 above)
Comment (tahb) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt:545]: It sounds good to me if you also think this type of thing would be of value.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt (hunk 2 above)
Comment (tahb) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:605]: I didn't know hat an infix function was! So the long hand is `"$.assessmentId".hasValidationError("doesNotExist")`. Thanks :)

**Pull request merged**